### DISCUSSION TITLE
How to effectively sync small number of folders

### DISCUSSION THREAD

**Discussion by LBrinston** [abraunegg/onedrive#2295] [Q&A] [answered]
I am attempting to sync a single directories (at the moment) out of my onedrive. I'm running Ubuntu 22.04 on a Microsoft surface (limited storage - hence not wanting to pull down my whole drive) for taking notes at University. 

My initial setup of onedrive seemed to work flawlessly but then I tweaked my sync_list (going down from two directories to one).

So I ran a `--resync --verbose` which has not ended. I've let it run up to an hour (I don't remember it taking this long on the initial setup) with it seemingly looping back over excluded files:

```
Skipping item - excluded by sync_list config: Documents/Unity 
Processing 220 OneDrive items to ensure consistent local state due to sync_list being used 
Skipping item - excluded by sync_list config: Documents/Unity
Processing 219 OneDrive items to ensure consistent local state due to sync_list being used
Skipping item - excluded by sync_list config: Documents/Unity
Processing 219 OneDrive items to ensure consistent local state due to sync_list being used
Skipping item - excluded by sync_list config: Documents/Unity
Processing 224 OneDrive items to ensure consistent local state due to sync_list being used
Skipping item - excluded by sync_list config: Documents/Unity
Skipping item - excluded by sync_list config: Documents/Education
Processing 220 OneDrive items to ensure consistent local state due to sync_list being used
Skipping item - excluded by sync_list config: Documents/Unity
Processing 219 OneDrive items to ensure consistent local state due to sync_list being used
Skipping item - excluded by sync_list config: Documents/Unity

```
Initially I thought the `--verbose` might be clipping the length of file paths but I also see the same Processing ###'s appear repeatedly.
Am I using the sync_list wrong? Should I be excluding unwanted directories as well? Am I misunderstanding how `--resync` works? Does it end?

I've perused the discussions for a similar issue and even removed/reinstalled onedrive only to experience the same seemingly endless `-resync`

**Comment by abraunegg** [comment, marked as answer]
@LBrinston 
Firstly, given that you are using ~Crapbuntu~ Ubuntu - please ensure that you have installed the client correctly as per https://github.com/abraunegg/onedrive/blob/master/docs/ubuntu-package-install.md

> Am I misunderstanding how --resync works? 

A `--resync` does end, however, if you are running the client from Ubuntu Universe - you probably are hitting a number of bugs related to sync looping that were fixed post 2.4.13 (which is the version that 22.04 includes .....) .... so see the above point about installing the correct client version.

A `--resync` should only *ever* be used when the client advises you to use it. Do not blindly always start the client with `--resync` - this unfortunately is a mistake that perpetuated in Ubuntu forums and discussions ...

> Does it end?

Yes it does end. What you are seeing is the client processing item sets from OneDrive . There is zero control as to how to tell OneDrive how many items to send, so they are sent in 200-240 bundle lots. Depending on how many files + folders you have - the number of item sets sent to you can be estimated: total files and folders online / 220 = number of `Processing X OneDrive items to ensure consistent` you will see.

**Reply by LBrinston** [reply]
I have removed and reinstalled as per the installation instructions for Ubuntu 20.04 and reauth'd my account.
Double checking I see:
```
loach@hippoglossus:~/.config/onedrive$ apt show onedrive -a
Package: onedrive
Version: 2.4.23-1+np1
Priority: optional
Section: net
Maintainer: Norbert Preining <norbert@preining.info>
Installed-Size: 2,821 kB
Depends: libc6 (>= 2.34), libgcc-s1 (>= 3.0), libglib2.0-0 (>= 2.12.0), libnotify4 (>= 0.7.3), libphobos2-ldc-shared98 (>= 1:1.28.0), libsqlite3-0 (>= 3.7.14), init-system-helpers (>= 1.52), libcurl4
Homepage: https://github.com/abraunegg/onedrive
Download-Size: 744 kB
APT-Manual-Installed: yes
APT-Sources: https://download.opensuse.org/repositories/home:/npreining:/debian-ubuntu-onedrive/xUbuntu_22.04 ./ Packages
Description: folder synchronization with OneDrive
 OneDrive is the cloud storage system of Microsoft. This package provides
 the command line client specialising in synchronizing with OneDrive cloud
 storage.
loach@hippoglossus:~/.config/onedrive$ apt list onedrive
Listing... Done
onedrive/unknown,now 2.4.23-1+np1 amd64 [installed]
onedrive/unknown 2.4.23-1+np1 arm64
onedrive/unknown 2.4.23-1+np1 armhf
```
I have thus far only used `--resync` when onedrive advises me to do so.

I appear to be hitting the same behavior again running:
`onedrive --synchronize --resync --dry-run --verbose`
BUT I did have to `Ctrl+c` it initially as I forgot to add `--verbose` and didn't want to sit around guessing if it was looping or not. I don't know if interrupting the `--resync` might be part of the issue? 

> A --resync should only ever be used when the client advises you to use it. Do not blindly always >start the client with --resync - this unfortunately is a mistake that perpetuated in Ubuntu forums > and discussions 

Should I be running `--synchronize` first thing before I setup the sync_list? Once I create the sync_list (the first thing I've been doing) I'm prompted to `--resync`.

**Comment by abraunegg** [comment]
@LBrinston 
It is best to configure 'sync_list' first, check your configuration with `--display-config` and then test with `--dry-run` to ensure that the client will do what you want.

From what you have written - it basically sounds like you are not giving the client time to process. When using `--resync` it is a way more intensive operation than without - you *have* to let it finish.

If you want to look at where the client is spending all it's time processing, you can add:
```
display_processing_time = "true"
```

To your 'config' file

**Reply by LBrinston** [reply]
Alright, I was clearly underestimating how long this operation should take. 
I have gotten the things `--resync`'d after the fresh install and then `--synchronize`'d. Both take about 40 minutes for 169786 files.

Just to confirm my understanding: 

There is no way around the sync taking that long (getting a wired internet connection aside) because this is purely client side filtering? The client has to parse all the files being sent by the server until it finds the files on the `sync_list`. 

This seems to be what I am observing after I changed a handful of files in the directory I wish to keep sync'd (the directory itself contains 9 sub-directories with 71 files mostly pdfs). I threw `--monitor` on and it dropped back into a full 40 minute sync. After that concluded I tested `--single-directory` and it also ran for 40 minutes.

**Comment by abraunegg** [comment]
@LBrinston 

Correct.

This is because `--resync` blows away your existing state.

Post the first sucessful sync after a resync, the state is known and further syncs can utilse what is known as the /delta sync point - meaning things are synced faster because the sync state is known.

Do not use `--resync` unless the application advises it. If you use resync you are back to going through all items online.

**Comment by abraunegg** [comment]
@LBrinston 
Further to this, when using `--monitor` the following process is performed:

1. At application startup an integrity check is performed of your data. This *may* or *may not* include a full scan of your data to ensure that your local data is correctly in sync
2. Post this initial startup scan, the client will drop into looping based on the following values:
  *  monitor_interval - https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#monitor_interval
  * monitor_fullscan_frequency - https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#monitor_fullscan_frequency
3. By default, a full scan of your data is performed once per hour
